[feature] implement infrastructure of calculating some simple metrics…

AmirhosseinZDev opened the pull request:
… of internal service calls in a main service call

PRD-116361

Commit:
[feature] implement infrastructure of calculating some simple metrics of internal service calls in a main service call
PRD-116361

minakh1993 commented on README.md:
````diff
@@ -300,6 +300,10 @@ you can add your desired exclude url patterns by creating HttpStatisticFilter be
 ```
+if you have a service that it calls some services internally ,and you need to log the statistics of these internal services you must annotate internal services method with `@Timer` and the filter logs the statistics of internal service calls like below:
+```
+ "-service" : "POST /httpserver/test", "internal service call duration" : [ { "serviceType" : "test3", "serviceName" : "test4", "duration" : 22 }], "total duration" : "23.128s", "active requests" : 0 } 
````
provide an example of how to use this annotation and then how it will result.

minakh1993 replied:
provide an example of how to use this annotation and then how it will result.

minakh1993 commented on tosan-httpserver-spring-boot-starter/src/main/java/com/tosan/http/server/starter/annotation/Timer.java:
```diff
@@ -0,0 +1,25 @@
+import java.lang.annotation.*;
+
+/**
+ * Annotation for calculate the execution statistics of internal service calls and log them
```
for calculating
remove and log them- instead use at next line:
statistics obtained from this annotation will be logged after request completion.

minakh1993 commented on tosan-httpserver-spring-boot-starter/src/main/java/com/tosan/http/server/starter/annotation/Timer.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Annotation for calculate the execution statistics of internal service calls and log them
+ * <br>
+ * if server has a service that it calls some other services internally and the statistics of these service call are needed ,
```
service calls

minakh1993 commented on tosan-httpserver-spring-boot-starter/src/main/java/com/tosan/http/server/starter/annotation/Timer.java:
```diff
@@ -0,0 +1,25 @@
+
+    /**
+     * @return Name of the webservice that is going to call internal service from it
+     */
```
return service type for categorizing purpose

minakh1993 commented on tosan-httpserver-spring-boot-starter/src/main/java/com/tosan/http/server/starter/annotation/Timer.java:
```diff
@@ -0,0 +1,25 @@
+    /**
+     * @return Name of the internal service that is going to call
+     */
+    String serviceName();
```
return service name

minakh1993 commented on tosan-httpserver-spring-boot-starter/src/main/java/com/tosan/http/server/starter/aspect/StatisticsAspect.java:
```diff
@@ -0,0 +1,38 @@
+ * @since 10/22/2022
+ */
+@Aspect
+@Order(Ordered.HIGHEST_PRECEDENCE)
```
why highest precedence?

AmirhosseinZDev replied:
because start time of service is calculated at the begining of service call and end time is calculated after all other aspects are proceed

minakh1993 commented on tosan-httpserver-spring-boot-starter/src/main/java/com/tosan/http/server/starter/aspect/StatisticsAspect.java:
```diff
@@ -0,0 +1,38 @@
+ */
+@Aspect
+@Order(Ordered.HIGHEST_PRECEDENCE)
+public class StatisticsAspect extends BaseAspect {
```
maybe TimerStatisticsAspect?

minakh1993 commented on tosan-httpserver-spring-boot-starter/src/main/java/com/tosan/http/server/starter/configuration/HttpServerUtilConfiguration.java:
```diff
@@ -124,4 +126,16 @@ public ServiceLoggingConfig serviceLoggingConfig() {
+
+    @Bean
+    @ConditionalOnMissingBean
+    public StatisticsAspect statisticsAspect(StatisticsUtil statisticsUtil) {
```
no config for aspect for now. so @ConditionalOnMissingBean does not seem necessary

minakh1993 commented on tosan-httpserver-spring-boot-starter/src/main/java/com/tosan/http/server/starter/configuration/HttpServerUtilConfiguration.java:
```diff
@@ -124,4 +126,16 @@ public ServiceLoggingConfig serviceLoggingConfig() {
+    }
+
+    @Bean
+    @ConditionalOnMissingBean
```
@ConditionalOnMissingBean not required for this util

minakh1993 commented on tosan-httpserver-spring-boot-starter/src/main/java/com/tosan/http/server/starter/filter/HttpStatisticsFilter.java:
```diff
@@ -81,9 +84,14 @@ private void end(String serviceName, long startTime) {
-        outwardLog.put("duration", duration + "s");
+        List<ServiceExecutionInfo> serviceExecutionInfos = Statistics.getCurrentStatistics().getServiceExecutionsInfo();
+        if (!serviceExecutionInfos.isEmpty()) {
+            outwardLog.put("internal service call duration", serviceExecutionInfos);
```
maybe at first more important data is better to be logged like total duration and active requests. log volume must be reduced as possible. so instead of internal service call duration we can go with statistics.
in logging service excecution info maybe we can go with serviceType.serviceName = duration....
consider that service type might be null.

cm sample of logging statistics:
2022-10-04T13:00:00.115 [INFO]  requestId:[NW-3mqkqJOo] channel:[10.5.35.123    ] thread:[qtp523794808-445527      ]  c.k.e.c.exporter.ServiceExporterBase.a@237:
    { "statistics" : "ePayment.ThirdPartyGetPaymentServices", "TrackerAspect" : "0.023s", "LoadSessionAspect" : "0.0s" } 
2022-10-04T13:00:00.115 [INFO]  requestId:[NW-3mqkqJOo] channel:[10.5.35.123    ] thread:[qtp523794808-445527      ]  c.k.e.c.exporter.ServiceExporterBase.a@212:
    { "-service" : "ePayment.ThirdPartyGetPaymentServices", "duration" : "0.066s", "operational active requests" : 2, "report active requests" : 0 }

minakh1993 commented on tosan-httpserver-spring-boot-starter/src/main/java/com/tosan/http/server/starter/statistics/Statistics.java:
```diff
@@ -0,0 +1,32 @@
+ * @since 10/22/2022
+ */
+public class Statistics {
+    private static final ThreadLocal<Statistics> currentStatistics = new ThreadLocal<>();
```
static final variable --> CURRENT_STATISTICS, and this thread local doesn't belong to just current statistics...

minakh1993 commented on README.md:
````diff
@@ -300,6 +300,10 @@ you can add your desired exclude url patterns by creating HttpStatisticFilter be
+```
+ "-service" : "POST /httpserver/test", "internal service call duration" : [ { "serviceType" : "test3", "serviceName" : "test4", "duration" : 22 }], "total duration" : "23.128s", "active requests" : 0 } 
+```
 
````
provide a sample of nested services with timer annotation in sample module

Commit:
[revision]d45bee1f
change some java doc, add a sample for Timer annotation, change format of internal statistics log

minakh1993 commented on README.md:
````diff
@@ -300,15 +302,31 @@ you can add your desired exclude url patterns by creating HttpStatisticFilter be
+public Object test4Service() {
+    ...
+}
+```
````
sample names:
 @Timer(serviceType = "InternalWebService", serviceName = "internalService")
    public void internalService() {

minakh1993 commented on tosan-httpserver-spring-boot-starter/src/main/java/com/tosan/http/server/starter/filter/HttpStatisticsFilter.java:
```diff
@@ -84,16 +84,33 @@ private void end(String serviceName, long startTime) {
+        List<String> statisticsDetail = new ArrayList<>();
+        for (ServiceExecutionInfo serviceExecutionInfo : serviceExecutionInfos) {
+            if (serviceExecutionInfo != null) {
+                if (!serviceExecutionInfo.getServiceType().isEmpty()) {
```
null pointer

minakh1993 commented on tosan-httpserver-spring-boot-starter/src/main/java/com/tosan/http/server/starter/filter/HttpStatisticsFilter.java:
```diff
@@ -84,16 +84,33 @@ private void end(String serviceName, long startTime) {
+                } else {
+                    statisticsDetail.add("-service : " + serviceExecutionInfo.getServiceName() + ", duration :" +
+                            serviceExecutionInfo.getDuration());
+
```
serviceName:duration

Commit:
[revision]d45bee1f
change README.md, change serviceType field of Timer too non-required field, change log format of statistics log

Commit:
[revision]d45bee1f
add unit test, add integration test
